Fallible only types

baumanj opened the pull request:
This is based on https://github.com/vcombey/fallible_collections/pull/3, so you can ignore the first commit for now and just look at https://github.com/vcombey/fallible_collections/commit/7e62b06ec26420c9f40cbe847830e96bff379ff1. I'll rebase after https://github.com/vcombey/fallible_collections/pull/3 is merged, but I wanted to put this here since I've already completed the work.

vcombey commented on src/vec.rs:
```diff
@@ -35,34 +38,433 @@ pub trait FallibleVec<T> {
     fn try_extend_from_slice(&mut self, other: &[T]) -> Result<(), TryReserveError>
     where
-        T: Copy + Clone;
+        T: Clone;
```
here the try_extend_from_slice function can panic on OOM if T is not copy and clone() fails
(because I have used `extend_from_slices` which "Iterates over the slice other, clones each element, and then appends it to this Vec").

baumanj replied:
Ah, good point. Will revert that bit.

vcombey commented on src/vec.rs:
```diff
@@ -35,34 +38,433 @@ pub trait FallibleVec<T> {
+        let mut v = Self::new();
+        v.extend_from_slice(value)?;
+        Ok(v)
+    }
```
Same here, why not require T: TryClone and use try_extend_from_slice_no_copy ?

vcombey commented on src/vec.rs:
```diff
@@ -35,34 +38,433 @@ pub trait FallibleVec<T> {
+        }
+        Ok(result)
+    }
+}
```
Same

vcombey commented on src/vec.rs:
```diff
@@ -35,34 +38,433 @@ pub trait FallibleVec<T> {
+    pub fn extend_from_slice(&mut self, other: &[T]) -> Result<(), TryReserveError> {
+        self.inner.try_extend_from_slice(other)
+    }
+}
```
Same

vcombey commented:
we could also implement TryClone for TryVec, TryBox... what do you think ?

baumanj commented:
> we could also implement TryClone for TryVec, TryBox... what do you think ?

Sounds good! Done in https://github.com/vcombey/fallible_collections/pull/4/commits/57ed9a3e82ab0fa1b5f625afa5428c1d75530f44

baumanj commented:
I think I've addressed all your feedback. I'll need to rebase this after https://github.com/vcombey/fallible_collections/pull/3 merges, but I wanted to wait on that until we're done reviewing the changes here since the force push required destroys the continuity of the review conversation.

baumanj commented:
I rebased this and added [some tests for `try_clone`](https://github.com/vcombey/fallible_collections/pull/4/commits/d7d4f8a907f1c1bd6a499fbfb58ae4359235b618). Once you merge this, can you publish a new version to crates.io? I think that's all I should need on your side to continue with https://bugzilla.mozilla.org/show_bug.cgi?id=1624698.
